lib/Utils/LLVMShorthands.cpp

```diff
@@ -164,12 +165,38 @@ std::string llvmIRToString(const llvm::Value *V) {
   return IRBuffer;
 }
 
+std::string llvmIRToStableString(const llvm::Value *V) {
+  std::string IRBuffer;
+  llvm::raw_string_ostream RSO(IRBuffer);
+  V->print(RSO, getModuleSlotTrackerFor(V));
+
+  boost::trim_left(IRBuffer);
+  RSO.flush();
+
+  if (auto Meta = IRBuffer.find_first_of("!#"); Meta != std::string::npos) {
+    IRBuffer.erase(Meta);
+
+    boost::trim_right(IRBuffer);
+    assert(!IRBuffer.empty());
+    if (IRBuffer.back() == ',') {
+      IRBuffer.pop_back();
+    }
+  }
```

Comment:
You could speed this up and simplify the code by using `llvm::StringRef`.
Do all the removal on the `StringRef` and, at the end, only once convert it back a string (before the append lines below).

Reply:
Working with StringRef has the problem of moving bytes inside the same buffer at the end, when we want to append the metadata Id. 
Of course, we could use `.str()` to create a copy of the StringRef, but I would like to avoid a further allocation. The same holds, when working with llvm::Twine.

Copying back the StringRef with std::memmove works, but does not update the IRBuffer's size.
Hence, I keep the in-place operations on IRBuffer for now as all but the `trim_left` are erasing from the end which should be fast.

If you know a better alternative, please let me know.

Reply:
Here is what I envisioned :laughing:
https://godbolt.org/z/qoPnTcGr9

The generated code is a bit longer but, on the other hand, all 'StringRef' modifications are inlined and don't need a function call to boost, and, overall, the code has less unnecessary constructions of, e.g., std::locale. And, as far as I can tell there is no extra allocation, just a memcpy for updating the IRBuffer.

However, I did not profile it, so I'm not sure if the difference does matter.

PS: With this comment, my intention is not that you neeeed to use the version without boost, I personally just like doing things in "pure" LLVM (if possible) without external library usage, as it turned out many times profitable for me. LLVM internals often work well together with other LLVM internals and, hence, can produce faster code by default.